### ISSUE TITLE
Some notifications include update and canonical link in links array

### ISSUE THREAD

**Issue by antje-s** [wmo-im/wis2-notification-message#82]
Should we add to Requirement 11:
"H": The links array property SHALL provide only one link with a link relation of http://def.wmo.int/def/rel/wnm/-/deletion, http://def.wmo.int/def/rel/wnm/-/update or IANA link relation of canonical"
to be more precise, whether it is new data, an update or a deletion.

**Comment by amilan17** [comment]
for reference, the other requirements in Req. 11: 

* E | For new data and metadata notifications, the links array property SHALL provide at least one link with an IANA link relation of canonical to clearly identify the preferred access link.
* F | For data or metadata update notifications, the links array property SHALL provide at least one link with a link relation of http://def.wmo.int/def/rel/wnm/-/update to clearly identify the preferred access link.
* G | For data or metadata deletions, the links array property SHALL provide at least one link with a link relation of http://def.wmo.int/def/rel/wnm/-/deletion to clearly identify data which has been deleted or removed.

**Comment by amilan17** [comment]
Note, this is different than what we are doing with WCMP2 which has these distinct properties instead: 
 
* created | optional | The date that the link was created
* updated | optional | The date that the link was updated

**Comment by amilan17** [comment]
@wmo-im/tt-wismd

**Comment by josusky** [comment]
@antje-s are you objecting the formulation "at least one" and would prefer to have there "exactly one"? 
I mean, it would make sense. Until now, we implicitly expected that one notification talks about one data (or metadata) item/record, not about multiple.  We have array of links, to be able to provide **additional** information (in the form of links). I do not think anyone really expects a notification with multiple links marked as "canonical" or as "http://def.wmo.int/def/rel/wnm/-/deletion".
@tomkralidis any comments from your side?

**Comment by antje-s** [comment]
The messages should not contain links for canonical and update (or deletion) at the same time. The notification is either for new data OR an update for already published data. Change to "exactly one" would be good, but then someone could still include "exactly one" canonical link AND "exactly one" update link.

**Comment by josusky** [comment]
I have also stumbled upon notifications with both "canonical" and "update" links. We should definitely clarify this. The current formulation allows multiple canonical links too. I suggest to change `/req/core/links` (Requirement 11) as follows:

> E. For new data and metadata notifications, the links array property SHALL provide exactly one link with an IANA link relation value `canonical` to clearly identify the preferred access link.
> F. For data or metadata update notifications, the links array property SHALL provide exactly one link with a link relation value `update` to clearly identify the preferred access link.
> G. For data or metadata deletions, the links array property SHALL provide exactly one link with a link relation value  `deletion` to clearly identify data which has been deleted or removed.

and add one more requirement:
> I. The links array property SHALL contain only one link with relation value `canonical`, `update`, or `deletion`. These relations cannot be combined within one WNM.

Of course, suggestions for better formulations are welcomed, but the current formulations with "at least one" are obviously causing confusion.

**Comment by tomkralidis** [comment]
TT-WISMD 2024-11-27
- option to update WNM per:
  - update Requirements 11E and 11F to update "at least one" to "exactly one"
  - add 11I per @josusky 's comment `The links array property SHALL contain **only one** link with relation value canonical, update, or deletion. These relations SHALL NOT be combined within one WNM.`

ACTION: @amilan17 to investigate update workflow

**Comment by tomkralidis** [comment]
TT-WISMD 2025-02-19
- need to assess impact of change and how to move it through the WMO process
- clarify w/ ET-W2IT
- needs analysis of existing WIS2 Nodes, and check w/ brokers/caches on how they deal with functionality
- specification needs clarification, we can give grace period for implementation

ACTION: @tomkralidis to discuss w/ ET-W2IT

**Comment by josusky** [comment]
Just for the record. JSON schema has concept of "oneOf" but there is no way we can use it to constrain link relation types, mainly because we allow all kind of links in the `links` array, not just `canonical`, `update` and `deletion`. So the validation of link `rel` attributes has to be done by some code (ETS).

**Comment by tomkralidis** [comment]
> Just for the record. JSON schema has concept of "oneOf" but there is no way we can use it to constrain link relation types, mainly because we allow all kind of links in the `links` array, not just `canonical`, `update` and `deletion`. So the validation of link `rel` attributes has to be done by some code (ETS).

Exactly.  If WNM link relations were bound to `canonical`, `update`, `deletion`, `oneOf` would have been a viable option.
